### PR TITLE
Erisch/features/devops/integration tests

### PR DESCRIPTION
**Pull request by erikschlegel**
## Test Automation Harness:
-------------------------------------
* [YES] Have you followed the guidelines in our Contributing [document](./CONTRIBUTING.md)?
* [YES] Have you added an explanation of what your changes do and why you'd like us to include them

**Description**
This pull request adds a new automated test harness to account for unit and integration tests. We recommend cobalt contributors provide automated test artifacts in the `test` folder  for each terraform template. We're using `mage` to orchestrate the test run. This harness supports integration test classes where thew class name starts with `TestIT` and unit tests that start with `TestUT`.

The test runner will automatically run integration and unit tests based on the git change log of the current working branch. We took a naive approach with re-running all template tests if TF module artifacts are modified. We'll follow up a with a PR to build a module reference <> template dependency graph to follow a more robust and sophisticated approach. 

We also put a focus on developer experience with simplifying local automated test runs with including a dockerized test runner. The dockerized test harness will be migrated to azure devops once the base image is multi staged. 

**Feature List**
- Integrating the test harness with our Azure DevOPS CI pipeline. 
- A dockerized test harness that can be invoked via `./test-harness/local-run.sh`
- Excluding unmodified terraform templates from test run
- An environment template file `.env.template`
- `dep` manifest for terratest dependencies. 
- Base docker image
- An nginx container hello world app service terraform template `azure-simple-hw`
- A backend template for setting up remote terraform state resources in azure. This template references the `backend-state` module from bedrock. 
- An golang based integration test for the hello world template. This test asserts that an http connection can be established to the app service resource.
- Removed unused files. 
- A shell script utility to automate the base image build step. 
- The `magefile.go` test orchestrator.
- Added a lint format step to our test runner. 
- Added key vault integration with our azure devops pipeline.

[NO] Does this include new/updated documentation?
**Details**
 To avoid tacking more changes onto this mounting PR I opted to submit a follow on PR that focuses on documentation. The user story will be closed once both PR's are merged. 

[YES] I have added tests to cover my changes.
[YES] All new and existing tests passed.
[YES] My code follows the code style of this project.
[YES] I ran lint checks locally prior to submission.
[YES] Have you checked to ensure there aren't other open [Pull Requests](../../pulls) for the same update/change?

Issue Number: #56 

## Does this introduce a breaking change?
-------------------------------------
- NO

## Any relevant logs, error output, etc?
-------------------------------------
[Logs](https://dev.azure.com/csedallascrew/project-cobalt/_build/results?buildId=159) from passing CI build from azure DevOPS

image

### CODE REVIEW

File: infra/templates/backend-state-setup/terraform.tfvars [hunk 1]
```diff
@@ -0,0 +1,2 @@
+name="cobaltfstate"
+location="<az-resource-group-location>"
```
Comment (nmiodice) [inline infra/templates/backend-state-setup/terraform.tfvars]: How does this value get inflated? Seems easy to miss and accidentally get this default string as our location.

File: infra/templates/backend-state-setup/variables.tf [hunk 2]
```diff
@@ -0,0 +1,17 @@
+variable "name" {
+  type        = "string"
+  description = "(Required) Specifies the human consumable label for this resource."
+  default     = ""
```
Comment (nmiodice) [inline infra/templates/backend-state-setup/variables.tf]: You may want to instead omit the default completely. This will cause the TF commands to fail unless its provided as an argument explicitly. 

similar comment for other variables in the project.

File: infra/templates/backend-state-setup/variables.tf [hunk 3]
```diff
@@ -0,0 +1,17 @@
...
+}
+
+variable "location" {
+  type        = "string"
+  description = " (Required) Specifies the supported Azure location where the resource exists. Changing this forces a new resource to be created."
```
Comment (nmiodice) [inline infra/templates/backend-state-setup/variables.tf]: required is redundant as the TF will fail with a similar note if its not provided as an arg

File: magefile.go [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+
+// The default target when the command executes `mage` in Cloud Shell
+var Default = Full
+
+func main() {
+	Default()
+}
+
+// A build step that runs Clean, Format, Unit and Integration in sequence
+func Full() {
```
Comment (nmiodice) [inline magefile.go]: nit: more clear if named something like `FullTestSuite` or `RunAllTests`

Similar -- `Unit` and `Integration` may be more clear if its `RunUnitTests` or `RunIntegrationTests`

... total nit. ignore if you want.

File: test-harness/build-base-image.sh [hunk 5]
```diff
@@ -0,0 +1,36 @@
+#!/usr/bin/env bash
+set -euo pipefail
+
+# Arguments
+declare go_version=""
+declare tf_version=""
+declare docker_img="msftcse/cobalt-test-base"
+declare docker_file="test-harness\docker\base-images\Dockerfile"
```
Comment (nmiodice) [inline test-harness/build-base-image.sh]: Why are we using `/` and `\`?

File: test-harness/Dockerfile [hunk 6]
```diff
@@ -0,0 +1,15 @@
+ARG base_img_tag
+
```
Comment (nmiodice) [inline test-harness/Dockerfile]: Can you add comments about the usage of this file vs `test-harness/docker/base-images/Dockerfile`?

File: test-harness/init.sh [hunk 7]
```diff
@@ -0,0 +1,66 @@
+#!/usr/bin/env bash
+set -euo pipefail
+
+declare -A TEST_RUN_MAP
+declare readonly TEMPLATE_DIR="infra/templates"
+
+function check_required_env_variables() {
```
Comment (nmiodice) [inline test-harness/init.sh]: could be shortened using a for loop that consumes a list of args

File: test-harness/init.sh [hunk 8]
```diff
@@ -0,0 +1,66 @@
+#!/usr/bin/env bash
```
Comment (nmiodice) [inline test-harness/init.sh]: Can you add a documentation here that explains the use of this file? I see what its doing, but its complicated enough that some guidance would help IMO

File: test-harness/local-run.sh [hunk 9]
```diff
@@ -0,0 +1,51 @@
+#!/usr/bin/env bash
```
Comment (nmiodice) [inline test-harness/local-run.sh]: Thank you for making local testing easier!!

Unsure how close we can get here, but what do you see as the absolute closest we can have our local testing match the testing from the pipeline steps? This is always a huge pain point that I have. If you could solve it (perhaps this approach already does) it would be a huge win IMO.

File: infra/templates/azure-simple-hw/test/azure_simple_integration_test.go [hunk 10]
```diff
@@ -0,0 +1,100 @@
+package test
```
Comment (nmiodice) [inline infra/templates/azure-simple-hw/test/azure_simple_integration_test.go]: This looks like a fair bit of code that is unrelated to the assertions you want to make. Have you considered introducing a layer of abstraction that could be used to make the tests standardized and easy to write?

File: Gopkg.toml [hunk 11]
```diff
@@ -0,0 +1,42 @@
+# Gopkg.toml example
```
[the remark this answers is not in the document]
Comment (erikschlegel) [reply Gopkg.toml]: I'm still a golang noob and not sure how much control we have on these files as they were auto generated.

File: devops/providers/azure-devops/templates/azure-pipelines.yml [hunk 12]
```diff
@@ -46,3 +68,37 @@ steps:
...
+- script: |
+    go run $TEST_HARNESS_GOFILE
+  env:
+    ARM_CLIENT_ID: $(AD-SP-CLIENT-ID)
+    ARM_CLIENT_SECRET: $(AD-SP-SECRET)
+    ARM_SUBSCRIPTION_ID: $(AD-SP-SUBSCRIPTION-ID)
+    ARM_TENANT_ID: $(AD-SP-TENANT-ID)
+    ARM_ACCESS_KEY: $(ARM-ACCESS-KEY)
+  workingDirectory: '$(modulePath)/$(BUILD_TEMPLATE_DIRS)'
+  displayName: 'Run Unit + Integration tests'
```
[the remark this answers is not in the document]
Comment (erikschlegel) [reply devops/providers/azure-devops/templates/azure-pipelines.yml]: This is a valid point. We could split things out where we run unit tests in the CI and integration tests in the deployment release. Running both unit and integration in the same CI matches the bedrock flow. Whichever we decide, we should be consistent across both projects. @NathanielRose thoughts?

File: go.mod [hunk 13]
```diff
@@ -0,0 +1,68 @@
+module cobalt
+
+require (
```
[the remark this answers is not in the document]
Comment (erikschlegel) [reply go.mod]: great question. This is auto pulled in through the terratest `imports`

```
	http_helper "github.com/gruntwork-io/terratest/modules/http-helper"
	random "github.com/gruntwork-io/terratest/modules/random"
	terraform "github.com/gruntwork-io/terratest/modules/terraform"
	test_structure "github.com/gruntwork-io/terratest/modules/test-structure"
	terraformCore "github.com/hashicorp/terraform/terraform"
```

File: infra/templates/azure-simple-hw/test/azure_simple_integration_test.go (hunk 10 above)
Comment (erikschlegel) [reply infra/templates/azure-simple-hw/test/azure_simple_integration_test.go]: that's a good point. As per our side chat this will be a top focus for our next sprint.

File: test-harness/local-run.sh (hunk 9 above)
Comment (erikschlegel) [reply test-harness/local-run.sh]: at a minimum, both `local-run` and `azure-pipelines` should use the same test harness container(that's coming soon with multi stage builds). I'd have to think about an effective way to locally replicate the branch naming convention azure devops follows.

File: .env.template [hunk 14]
```diff
@@ -0,0 +1,16 @@
+BUILD_BUILDID=1
+BUILD_UPSTREAMBRANCH=master
+BUILD_SOURCEBRANCHNAME=`git branch | sed -n '/\* /s///p'`
```
[the remark this answers is not in the document]
Comment (erikschlegel) [reply .env.template]: that's a great point. I'll move those to `local-run.sh`.

File: devops/providers/azure-devops/templates/azure-pipelines.yml (hunk 12 above)
[the remark this answers is not in the document]
Comment (nmiodice) [reply devops/providers/azure-devops/templates/azure-pipelines.yml]: As per the discussion we had today I'm OK leaving this as is. It seems like there will be future work to extend and standardized the testing patterns between the project as more requirements are gathered.

File: magefile.go [hunk 15]
```diff
@@ -0,0 +1,71 @@
...
+// A build step that runs integration tests
+func RunIntegrationTests() error {
+	mg.Deps(CleanAll)
+	mg.Deps(FormatGolangFiles)
+	fmt.Println("INFO: Running integration tests...")
+	return sh.RunV("go", "test", "./...", "-run", "TestIT", "-v", "-timeout", "99999s")
+}
+
+// A build step that formats both Terraform code and Go code
+func FormatGolangFiles() error {
```
Comment (nmiodice) [inline magefile.go]: [question for the future] Should this be caught by linting and cause a build to fail? I think you may be able to run something like this to check if the terraform is correctly formatted:

```
# https://www.terraform.io/docs/commands/fmt.html
terraform fmt -check=true
```

File: infra/templates/azure-simple-hw/test/azure_simple_integration_test.go [hunk 16]
```diff
@@ -0,0 +1,113 @@
...
+		BackendConfig: map[string]interface{}{
+			"storage_account_name": tfstate_storage_account,
+			"container_name":       tf_state_container,
+		},
+	}
+
+	return terraformOptions
+}
+
+const SKIP_STAGE_ENV_VAR_PREFIX = "SKIP_"
```
Comment (nmiodice) [inline infra/templates/azure-simple-hw/test/azure_simple_integration_test.go]: What is the use case for this feature?

File: infra/templates/backend-state-setup/terraform.tfvars [hunk 17]
```diff
@@ -0,0 +1,2 @@
+name="cobaltfstate"
```
Comment (nmiodice) [inline infra/templates/backend-state-setup/terraform.tfvars]: As per our discussion, are we planning on preferring `variables.tf` or CLI args to `*.tfvars`?

File: magefile.go (hunk 15 above)
Comment (erikschlegel) [reply magefile.go]: Agree that we should create a user story to support a proper TF Lint target within our test harness.

File: devops/providers/azure-devops/templates/azure-pipelines.yml [hunk 18]
```diff
@@ -20,29 +22,64 @@ pr:
...
 
 steps:
+- checkout: self
+  persistCredentials: true
+  clean: true
+
+- script: |
+    docker login $ACR_HOST -u $ACR_USERNAME -p $(DOCKER-PASSWORD)
+
+- script: |
```
Comment (NathanielRose) [inline devops/providers/azure-devops/templates/azure-pipelines.yml]: So checked with AzDO there isnt a way to trigger tasks from repo path and this is a good approach to acheive this. Another approach would have been to have multiple build jobs based on repo file path  that would execute different parts of the test suite. Only concern here is it may get tricky to get the logs out running everything inside docker.
